### PR TITLE
Add TokenCredentials for Azure Event Hub

### PR DESCRIPTION
**Pull request by marioleed**
**What this PR does / why we need it**:
Add TokenCredentials overloads for AddAzureEventHub to simplify the use of MSI. 

**Which issue(s) this PR fixes**:

Please reference the issue this PR will close: #1537 

**Special notes for your reviewer**:
You could still use `EventHubConnection` but I'm not sure how well it works with a connection string in its current form. Looking at the code, if the user is using the `AzureEventHubHealthCheck(EventHubConnection connection)` constructor and `AzureEventHubHealthCheck.CheckHealthAsync()` tries to add a new `EventHubProducerClient` (in case the key doesn't exist - which shouldn't happen actually) it will use a connection string without a shared access signature key.

This is a way to get around that.

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [ ] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs [hunk 1]
```diff
@@ -9,55 +10,72 @@ public class AzureEventHubHealthCheck : IHealthCheck
...
 
-            _eventHubConnectionString = $"{connection.FullyQualifiedNamespace};{ENTITY_PATH_SEGMENT}{connection.EventHubName}";
-
-            if (!_eventHubConnections.ContainsKey(_eventHubConnectionString))
-            {
-                _eventHubConnections.TryAdd(_eventHubConnectionString, new EventHubProducerClient(connection));
+            string key = ConnectionKey;
+            if (!_clientConnections.ContainsKey(key))
+            {
+                _clientConnections.TryAdd(key, new EventHubProducerClient(_endpoint, _eventHubName, _tokenCredential));
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs:59]: `TryAdd` _may_ return `false`. I suggest you to look into `internal static class ClientCache` under `HealthChecks.CosmosDb` namespace.

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs [hunk 2]
```diff
@@ -9,55 +10,72 @@ public class AzureEventHubHealthCheck : IHealthCheck
...
         public AzureEventHubHealthCheck(EventHubConnection connection)
         {
-            Guard.ThrowIfNull(connection);
+            Guard.ThrowIfNull(connection);
+
+            _connectionString = $"{connection.FullyQualifiedNamespace};{ENTITY_PATH_SEGMENT}{connection.EventHubName}";
+
+            if (!_clientConnections.ContainsKey(_connectionString))
+            {
+                _clientConnections.TryAdd(_connectionString, new EventHubProducerClient(connection));
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs:46]: The same issue with `TryAdd` here.

File: test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureEventHubUnitWithTokenRegistrationTests.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            var services = new ServiceCollection();
+            services.AddHealthChecks()
+                .AddAzureEventHub(string.Empty, string.Empty, new AzureCliCredential());
+
+            using var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetRequiredService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+
+            Assert.Throws<ArgumentNullException>(() => registration.Factory(serviceProvider));
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureEventHubUnitWithTokenRegistrationTests.cs:125]: `Should.Throw<>` + check argument name

Comment (marioleed) [reply test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureEventHubUnitWithTokenRegistrationTests.cs:125]: Should I fix this in the AzureServiceBus* tests as well or should we leave it for later?

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs (hunk 1 above)
Comment (marioleed) [reply src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs]: Added ClientCache 👍🏻

File: test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureEventHubUnitWithTokenRegistrationTests.cs (hunk 3 above)
Comment (sungam3r) [reply test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureEventHubUnitWithTokenRegistrationTests.cs]: Up to you.

File: src/HealthChecks.AzureServiceBus/ClientCache.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+using System.Collections.Concurrent;
+
+namespace HealthChecks.AzureServiceBus;
+
+internal static class ClientCache
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/ClientCache.cs:5]: I suggest to move it into source-shared class (from CosmosDb) + remove another copy from HealthChecks.AzureStorage. Example of source-shared class is `<Compile Include="../CallerArgumentExpressionAttribute.cs" Link="CallerArgumentExpressionAttribute.cs" />` in Directory.Build.props. You may include this file under condition by `MSBuildProjectName` (if you like) to not clutter other projects.

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs [hunk 5]
```diff
@@ -1,63 +1,69 @@
...
 
-            _eventHubConnectionString = connectionString.Contains(ENTITY_PATH_SEGMENT) ? connectionString : $"{connectionString};{ENTITY_PATH_SEGMENT}{eventHubName}";
+        _connectionString = $"{connection.FullyQualifiedNamespace};{ENTITY_PATH_SEGMENT}{connection.EventHubName}";
 
-            if (!_eventHubConnections.ContainsKey(_eventHubConnectionString))
-            {
-                _eventHubConnections.TryAdd(_eventHubConnectionString, new EventHubProducerClient(_eventHubConnectionString));
-            }
-        }
+        ClientCache.GetOrAdd(ConnectionKey, _ => new EventHubProducerClient(connection));
```
[the remark this answers is not in the document]
Comment (marioleed) [reply src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs:38]: Initially I was thinking of scoping this PR to only include EventHub stuff as to not make such a big impact. 

With that said, I was also considering `ClientCache.GetOrAddDisposableAsync` as well. But then thought I wouldn't need it as async methods shouldn't be called from the constructor (which would happen in this case). I guess the problem is the `EventHubConnection` constructor which will not work properly in it's current form with a connection string. I'll try to move it to a field instead. Let's see what I can do 🙂

File: src/HealthChecks.AzureServiceBus/ClientCache.cs (hunk 4 above)
Comment (marioleed) [reply src/HealthChecks.AzureServiceBus/ClientCache.cs]: Moved it and scoped it in Directory.Build.props to the three projects using ClientCache, AzureServiceBus, AzureStorage and CosmosDb. However, I have not changed to any async usages in AzureStorage or CosmosDb. Nor have I added ClientCache in the AzureServiceBus* classes. Don't know if I should or not, but I'm fine with that being a new PR 😄

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs [hunk 6]
```diff
@@ -1,63 +1,73 @@
...
-
-            _eventHubConnectionString = $"{connection.FullyQualifiedNamespace};{ENTITY_PATH_SEGMENT}{connection.EventHubName}";
-
-            if (!_eventHubConnections.ContainsKey(_eventHubConnectionString))
-            {
-                _eventHubConnections.TryAdd(_eventHubConnectionString, new EventHubProducerClient(connection));
-            }
-        }
+            var client = await ClientCache.GetOrAddAsyncDisposableAsync(ConnectionKey,
+                async _ => await Task.FromResult(CreateClient()));
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs:51]: I would not juggle async/await here. Just add another ` public static async ValueTask<T> GetOrAddAsyncDisposableAsync<T>(string key, Func<string, T> clientFactory) where T : IAsyncDisposable` overload and use it here. `IAsyncDisposable` is about asyncronous destroying, not about asyncronous creation.

Comment (marioleed) [reply src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs]: That sounds more sane 😄 Done.

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
Forgot to bump versions in `dependencies.props`.

**Comment by marioleed** [comment]
> Forgot to bump versions in `dependencies.props`.

Ah ok. Yeah that seems interesting. I guess it's mostly minor changes. Otherwise, do advice.

File: build/dependencies.props [hunk 7]
```diff
@@ -14,11 +14,11 @@
     <HealthCheckAWSSqs>6.0.0</HealthCheckAWSSqs>
     <HealthCheckAWSSystemsManager>6.0.0</HealthCheckAWSSystemsManager>
     <HealthCheckAzureDigitalTwin>6.0.0</HealthCheckAzureDigitalTwin>
-    <HealthCheckAzureServiceBus>6.0.4</HealthCheckAzureServiceBus>
-    <HealthCheckAzureStorage>6.1.1</HealthCheckAzureStorage>
+    <HealthCheckAzureServiceBus>6.1.0</HealthCheckAzureServiceBus>
```
Comment (marioleed) [inline build/dependencies.props:17]: @sungam3r Does these bumps look okay?

Comment (sungam3r) [reply build/dependencies.props:17]: Yes. Look into approval files. No changes - bump patch. Only green lines - bump minor. Else bump major.

**Pull request merged**